### PR TITLE
kafka: add stub classes for mesh filter

### PR DESCRIPTION
**Pull request by adamkotwasinski**
Commit Message: kafka: add stub classes for mesh filter
Additional Description: stub for Kafka mesh filter
creates a new filter object, abstraction for in-flight request, very simple request processor (that doesn't process any of requests); provides generic logic of how the mesh will work on a high level:
* receive the kafka request
* run it thru decoder (this is existing code from broker-filter)
* run decoded message thru processor to enrich it into actionable object
* processor brings it back to filter to store it and start processing
* (not in this PR) depending on request type, we engage upstream brokers (we finally send records) or are ready immediately
* (not in this PR) request notifies owner (filter) that it's finished and owner (filter) can finally respond on connection

Risk Level: low, code cannot be referenced as it doesn't have an extension
Testing: unit tests right now
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

**Comment by adamkotwasinski** [comment]
Let's run builds for now, most probably something will be red and I'll notice some more places to add comments.

File: source/extensions/filters/network/kafka/mesh/abstract_command.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+namespace Mesh {
+
+/**
+ * Represents single downstream client request.
+ * Responsible for performing the work on multiple upstream clusters and aggregating the results.
+ */
+class InFlightRequest {
```
Comment (adamkotwasinski) [inline source/extensions/filters/network/kafka/mesh/abstract_command.h:16]: implementations of this interface will provide the proper handling to provide kafka-mesh features
to have basic compliance with protocol we will require to handle the following types
* api versions - [to tell client we understand some requests](https://github.com/envoyproxy/envoy/pull/11936/files#diff-1c8b61805997d18486e049aa302713356f53a6fc53eb6ecd6ba29fdbf41aca1bR29) 
* metadata - [to tell client to send produce requests to our envoy instance](https://github.com/envoyproxy/envoy/pull/11936/files#diff-6f031aa07a0ebd92856b2d409255e4a043842edb60a28cb5d8f916c4e78e110aR27)
* produce - [to actually process the data and send them upstream to kafka](https://github.com/envoyproxy/envoy/pull/11936/files#diff-a5b85c101e48a968f816536f17e9e7ab84db7abf19e468826f30085260ad2421R23)

File: source/extensions/filters/network/kafka/mesh/request_processor.cc [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+// Helper function. Throws a nice message. Filter will react by closing the connection.
+static void throwOnUnsupportedRequest(const std::string& reason, const RequestHeader& header) {
+  throw EnvoyException(absl::StrCat(reason, " Kafka request (key=", header.api_key_, ", version=",
+                                    header.api_version_, ", cid=", header.correlation_id_));
+}
+
+void RequestProcessor::onMessage(AbstractRequestSharedPtr arg) {
+  // This will be replaced with switch on header's API key.
+  throwOnUnsupportedRequest("unsupported (bad client API invoked?)", arg->request_header_);
```
Comment (adamkotwasinski) [inline source/extensions/filters/network/kafka/mesh/request_processor.cc:19]: when `arg` can be actually turned into instance of [InFlightRequest](https://github.com/envoyproxy/envoy/pull/17374/files#diff-bc27f6cc82b6e0abc63f74ed16db37d6b3319fed85d487fecf68d639de5ba287R16) then we can [go back to filter](https://github.com/envoyproxy/envoy/pull/17374/files#diff-54af050994ce6e17b4f69b01159b9ba08721f38cb4698d6ad717cf6b3ee03626R57)

**Review by mattklein123 — APPROVED** [review]
Thanks. Since this extension is going to get moved to contrib I quickly skimmed and am going to merge.

**Pull request merged**